### PR TITLE
Add NAP for multicanvas

### PR DESCRIPTION
**Pull request by aganders3**
Particular thanks to @brisvag and @melonora for previous work and discussions on this, and for agreeing to be co-authors. I am happy to absorb any criticism and direct praise their way 😄.

Thanks also to @andy-sweet and @Czaki, both of whom provided helpful discussion while I worked on this NAP.

Of course I expect many changes to this and look forward to feedback.

### CODE REVIEW

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: Would this benefit the work here or do you think of it as mostly orthogonal? Would a simultaneous effort help or hinder progress here?

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: One alternative here that came to mind is that each canvas (or maybe vispy canvas) could own its own layer slicer - that way you don't need to pass in the extra canvas parameter and you can do cancellation per canvas (which I think is desirable).

With lots of canvas, you might introduce too much concurrency, but in most cases, maybe that won't be too bad. And we may be able to serialize it with a shared resource/controller if needed.

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: Maybe I misread the out-of-scope section, but my impression was that moving slice state off the layer was a non-goal - is that wrong?

Or maybe this is the proposed approach that incidentally achieves that non-goal?

Anyway, I'm generally in favor of that move, but just wanted to clarify the proposal and intent here.

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: I'm fully in support of this ordering where implementation of architecture changes happen first to support new functionality later.

One mistake on the async slicing project is that I didn't do enough to prioritize things that way, which meant that things didn't end up quite as I'd hoped.

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: Typo?

```suggestion
    * through-plane slice indicators
```

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: I think that shouldn't be needed and if it is, I'm happy to help make that need go away.

Also, all the typos :D

```suggestion
[^async-only]: Depending on the timeline and prototype implementation, it may be acceptable/preferable for multi-canvas feature to rely on (currently experimental) async slicing ([see NAP-4](https://napari.org/stable/naps/4-async-slicing.html)).
```

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: Is there any prototype code available to look at and play with? And/or some videos showing some of the proposals here at least kind-of in action?

Both could help to ground some of the proposals and discussion here.

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [inline docs/naps/8-multiple-canvases.md]: I regretted adding something like this to the async slicing NAP, because it's just too much detail, so another option might be to remove it from the main NAP document and maybe link to it or shove it into an appendix.

**Review by andy-sweet — COMMENTED** [review]
I took a first pass through this and left a few comments. Overall, I'm very excited to see something like this get approval and get implemented!

File: docs/naps/8-multiple-canvases.md
Comment (melonora) [reply docs/naps/8-multiple-canvases.md]: I think layergroups here could potentially provide a cleaner API, but overall the API is similar to current grid mode at least for multichannel view, thus could be an orthogonal effort.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: Yeesh those typos 🙈

I agree I don't think this is necessary. This was speculation before I started prototyping but so far it hasn't been an issue. I can remove this note.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: I don't think it's entirely orthogonal, but perhaps mostly. I think it depends on the scope of features we want. For example just below I propose for all canvases to share a common layer list, with independent slicing, camera, dimensionality, and layer visibility. I think this would be more relevant if supporting (for example) different colormaps for the same data in two canvases.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: Thanks - that's an interesting idea I hadn't considered. I was initially thinking the current LayerSlicer may expand to use a pool instead of a single thread.

I think you're right passing the full canvas is probably more than necessary, and trying to pass a canvas ID or something feels a bit hacky to me. One benefit of just passing the canvas object is that it would bring with it the camera, which might be useful for finding `corner_pixels` without waiting for the draw.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: Do you mean this non-goal?
> Normalizing slice data for different layer types, though this may benefit in the course of this work.

That's not really what I meant there, but also I think it's ambiguous enough warrant changing or removing. I was trying to say it is a non-goal to create a uniform/generic concept (or Protocol) of a "slice" - it's okay for the slices to be quite specific for each layer type.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: I found it helpful to write this out, but if it's too detailed I'm happy to move or remove it. It might just fit better in a PR with actual code changes.

**Comment by aganders3** [comment]
> I took a first pass through this and left a few comments. Overall, I'm very excited to see something like this get approval and get implemented!

Thanks! I appreciate the support and feedback.

**Review by Czaki — DISMISSED** [review]
The number 8 is for telemetry (#226), also please add a new NAP to the table of contents.

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: My prototype/experimentation code is pretty early, but is here:
https://github.com/aganders3/napari/pull/13

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md:30]: I would make an extra effort from the get-go to define terminology:
- viewbox: specific concept in vispy, different from canvas
- viewer: I would specify "currently almost 1to1 matching to a window"

Opening multiple viewers is not only necessary for seeing the same data, but also for seeing different data, which does not incur the same memory complications and should be even easier, but is currently also impossible inside a single viewer.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: I would add the use cases with non-identical data, like viewing different segmentations overlaid to the same image, side by side.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: Grid mode is orthogonal to this effort, IMO, at least as we intended it in that issue; it should be drop-in with the current grid mode, with no public API changes, at least at first. What do you think would be useful from the grid mode efforts to the "true multicanvas"?

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: I guess these are numbers just to have some goal? I don't see why we would go for specificallty 8 other than "arbitrary"...

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md:66]: This is my only disagreement with the current form of the NAP, I wrote more details in the general review comment.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: As mentioned above, I would move this higher up for clarity. I would also make sure to distinguish what's the *current* definition versus this NAP's *preferred* definition. I know it's semantics, but this stuff is confusing (as evidenced from previous conversations, such as the one in https://github.com/napari/napari/pull/4684.)

For example, we really need to clarify that when this nap says "Canvas", it's *not* a vispy/opengl/other-backend `Canvas`, but a more complex object that's much closer to what the `ViewerModel` is now.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: For us non-IT people, please use the full name once for "MVC" before the acronym :P

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md:112]: Not sure I get this: `layer_to_slice` is supposed to be just a dict to keep track of which layer each slice in the canvasmodel comes from? If so, why is this necessary as opposed to the current event system? Or is it just a way to hold state about the sliced data *somehwere*? (in which case, my first instinct would be to hold sliced data *not* in the canvas, but higher up in the viewer (since the same layer might be present in different canvases).

EDIT: after reading below about the LayerSlicer updates, I'm more and more convinced that these slices should be held higher up for canvases to *take* from when necessary.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: I don't see any way/reason to have selection be different from one canvas to the other, so that seems like a pretty layer-tied concept rather than slice-tied. If one needs different selections at the same time on the same layer, they probably need two layers anyways?

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md:174]: This is hard to do unless we assume a shared space between canvases, which I'm not sure we should do. Initially, I think we should leave this to the user.

Creating an orthoview example as part of this would would be a great way to test these ideas.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]: Which changes? I don't think grid mode is related to any of the above bullet points (at least as we intended it in the referenced issue: multiple viewboxes inside a single vispy canvas, sharing dims/layerlist/..., linked cameras, and just spreading the layers across viewboxes with the same logic as the current grid). If you mean a different kind of Grid Mode, please expand on it here for clarity.

File: docs/naps/9-multiple-canvases.md
Comment (brisvag) [inline docs/naps/9-multiple-canvases.md]:
```suggestion
Maintaining the proxy API on the viewer via the concepts of a main and/or active canvas will make this work mostly backward-compatible, though there will inevitably be some breaking changes. There will likely be significant breaking changes to private APIs. For example if plugins are attempting to access slice data directly from a layer instance, it may no longer be as expected. If this is a large burden, it too may be mitigated by delegating from the layer to the slice corresponding to the main or active canvas.
```

**Review by brisvag — APPROVED** [review]
Generally, great work on the NAP! Other than comments and clarifications here and there, I have one main disagreement, and it seems it comes from a misunderstanding when we discussed before, because I thing this is quite important. You say:

```
* All canvases shall share a common layer list and (unsliced) layer data.
```

I think we need to ensure from the get-go that canvases can hold different layer lists, but can also share some layers. This is key for example for applications like highly multiplexed data (@melonora can add some more specifics here if needed). I fear that leaving this out will make it harder to support it afterwards.

Anyways, approving as draft as per normal NAP procedure :)

**Comment by aganders3** [comment]
Thanks for the review and great comments @brisvag! I will try to respond to each and update accordingly, but generally your comments cleared up a few things for me. We're not too far apart on anything here :).

File: docs/naps/8-multiple-canvases.md
Comment (andy-sweet) [reply docs/naps/8-multiple-canvases.md]: No, I meant this one.

> Separation of rendering information from the Layer(Data) model

But maybe you meant that only with respect to NAP-3.

Anyway, I also agree that a generic slice class/protocol should be a non-goal. Though we might consider during/after.

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md]: I took a look back at the discussion of improved grid mode, and I agree. It _could_ be implemented with multiple canvases, but I think it makes sense to implement with vispy viewboxes. I've just removed the references here.

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md]: Thanks - this makes sense. Removed!

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md]: Yes sorry!

**Comment by aganders3** [comment]
I updated a bit but not ready for your re-review @brisvag. I should be able to finish addressing your comments next week, but you can ignore GitHub notifications for now :)

File: docs/naps/8-multiple-canvases.md
Comment (aganders3) [reply docs/naps/8-multiple-canvases.md]: I see - I will try to make this more clear. The goal here is indeed to move the slice state off the layer. The non-goal was meant to indicate that the remaining attributes of the layer will not be separated. So for example this work does not necessarily make it easier to share data between two layers with different colormaps.

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md:112]: I think you're correct about this - I'm going to remove this and we can revisit if necessary. I agree it makes  more sense for slices to be held somewhere where they could potentially be grabbed by more than one canvas.

Actually I will hold off making specific changes around this until we resolve the other points around shared/independent layer lists. It's not entirely related but I will adjust the text and diagrams for this once that is settled.

**Comment by aganders3** [comment]
> I have one main disagreement, and it seems it comes from a misunderstanding when we discussed before, because I thing this is quite important. You say:
> > * All canvases shall share a common layer list and (unsliced) layer data.
> 
> I think we need to ensure from the get-go that canvases can hold different layer lists, but can also share some layers. This is key for example for applications like highly multiplexed data (@melonora can add some more specifics here if needed). I fear that leaving this out will make it harder to support it afterwards.

My way of addressing this use case was specifying layers would gain a per-canvas visibility setting. I think this is also where misalignment on https://github.com/napari/docs/pull/249#discussion_r1365484821 comes from; using a common layer list would also inherently share space between canvases. If you want a truly independent space, I don't see that much benefit over opening a separate viewer window.

Changing to fully independent layer lists would definitely offer more flexibility, and may actually make this implementation simpler in some ways. My concern is that it will make for a more difficult UI design, and may make it more difficult to implement certain common use cases later (for example, where you do want a common space between canvases). That said, I am definitely still open to this approach. You (and @melonora) may have a better understanding of the use-cases that would benefit.

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md]: Yes I think this was just to have a goal. I could see setting some limit to ensure a good experience, but that could probably be part of the UI/UX layer instead. I've removed this for now.

File: docs/naps/9-multiple-canvases.md
Comment (aganders3) [reply docs/naps/9-multiple-canvases.md:30]: I moved the definitions up and tried to clear this up. I agree it's important thought so let me know if you have more suggestions!

**Pull request merged**
